src/FocusEvent.ts

```diff
@@ -88,32 +86,45 @@ export function setupFocusEvent(win: Window): void {
 
   kwin.HTMLElement.prototype.focus = focus;
 
-  const data: KeyborgFocusEventData = (kwin.__keyborgData = {
-    focusInHandler: (e: FocusEvent) => {
-      const target = e.target as HTMLElement;
-      if (!target) {
-        return;
-      }
+  const focusInHandler = (e: FocusEvent) => {
+    let target = e.target as HTMLElement;
+    if (!target) {
+      return;
+    }
 
-      const event = document.createEvent("HTMLEvents") as KeyborgFocusInEvent;
+    if (target.shadowRoot) {
+      // https://bugs.chromium.org/p/chromium/issues/detail?id=1512028
+      // focusin events don't bubble up through an open shadow root once focus is inside
+      // once focus moves into a shadow root - we drop the same focusin handler there
+      // keyborg's custom event will still bubble up since it is composed
+      target.shadowRoot.removeEventListener("focusin", focusInHandler);
+      target.shadowRoot.addEventListener("focusin", focusInHandler);
```

Comment:
Is it really correct?  remove and add the same on the same target? How do we clean it up? Wouldn't it become a memory leak?

Reply:
`removeEventListener` is called so that we don't constantly add the same event listener to the target.

I thought that once DOM elements were removed, the browser cleaned up its event listeners. It's possible this might not be the case

Reply:
Does adding a shadow root event listener from focus in event of the parent mean that first focus in inside shadow root will not reach the target?

Reply:
The first focusin of the shadow root will reach the parent handler - the subsequent onces inside the shadow root will not - please read the chrome bug in the comments

Reply:
@mshoho  https://github.com/microsoft/keyborg/pull/51/commits/f3b525f69bea82b624808ff82b960dd91aa73557 Adds a `focusout` handler to shadow DOM so that once focus leaves the shadow DOM, the event listeners are cleaned up.

Reply:
I suppose let's merge it and try a new version out.